### PR TITLE
adding in ledger support

### PR DESCRIPTION
**Pull request by whyrusleeping**
Still some work to do here, but this does work! 

Basic instructions:

First build lotus-wallet (patched to use the ledger wallet code), configure your lotus daemon to use it, and run it.
Then,

```
lotus-shed ledger list #pick a path
```
```
lotus-shed ledger key-info "m/44'/461'/0'/0/0"
```

put the last line of that output into a file, `ledgerkey.json` or something
Start `lotus-wallet`:
```
lotus-wallet start --ledger
```
Setup config in lotus for remove wallet: `$LOTUS_PATH/config.toml`:
```
[Wallet]
RemoteBackend="http://127.0.0.1:1777" #address can be changed when running `lotus-wallet start`
```

Import the key info into the wallet:
```
lotus wallet import --format=json-lotus ledgerkey.json
```

Now the key should show up in `lotus wallet list`

If you try to sign anything with the key, for example:
```
lotus send --from=LEDGERADDR f010101 123
```

It will prompt you to approve the transaction on your ledger, blocking until you approve it.

### CODE REVIEW

File: chain/messagepool/messagepool.go [hunk 1]
```diff
@@ -1249,6 +1249,9 @@ func (mp *MessagePool) Updates(ctx context.Context) (<-chan api.MpoolUpdate, err
 }
 
 func (mp *MessagePool) loadLocal() error {
+	return nil
+	// TODO: this causes super slow startup...
```
Comment (magik6k) [inline chain/messagepool/messagepool.go:1253]: TODO: Remove before merge (Tho I guess we can also fix this while here)

File: cmd/lotus-shed/ledger.go [hunk 2]
```diff
@@ -27,14 +31,37 @@ var ledgerCmd = &cli.Command{
...
+			defer closer()
+		}
+		ctx := lcli.ReqContext(cctx)
 
 		fl, err := ledgerfil.FindLedgerFilecoinApp()
 		if err != nil {
 			return err
 		}
 
 		for i := 0; i < 20; i++ {
```
Comment (Kubuxu) [inline cmd/lotus-shed/ledger.go:60]: When iterating while looking at the balance, we should iterate 20 keys in front of the last account with balance. (BIP44)

File: chain/wallet/multi.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+	}
+	if w == nil {
+		return nil, xerrors.Errorf("key not found")
+	}
+
+	return w.WalletExport(ctx, address)
+}
+
+func (m MultiWallet) WalletImport(ctx context.Context, info *types.KeyInfo) (address.Address, error) {
+	w := firstNonNil(m.Remote, m.Ledger, m.Local)
```
Comment (Kubuxu) [inline chain/wallet/multi.go:136]: This means that you cannot import to a normal key if you have ledger, or ledger key if you have Remote (which I think might be fine).

Probably the solution is to make the Ledger key be different type of key.

**Review by magik6k — APPROVED** [review]

**Review by Kubuxu — APPROVED** [review]

**Comment by Kubuxu** [comment]
It would be great to write some docs on this.

**Comment by prastut** [comment]
Hi @Kubuxu I am working on a doc for this.

Also lotus-wallet start --ledger seems to not work. lotus-wallet run --ledger seems to be working. 

======================

Just wanted to see whether you guys face a problem with nonce? For some reason, any send command I send using the above setup and the nonce seems to be 0

Context for the full problem is posted on FIL Slack: 
https://filecoinproject.slack.com/archives/CPFTWMY7N/p1613314781222700

**Pull request merged**